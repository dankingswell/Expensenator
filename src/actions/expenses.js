import uuid from "uuid"
//Actions
// Add, remove,expense

//Uses defualt object for params

export const addExpense = ({description ="",
                            note = "" ,
                            amount=0,
                            createdAt=0  
                                } ={}) => 
    ({
    type:"ADD_EXPENSE",
    expenses: {
        id:uuid(),
        description,
        note,
        createdAt,
        amount
    }
})

export const editExpense = ( id, updates) => (
    {
        type:"EDIT_EXPENSE",
        id,
        updates
    }
)

export const removeExpense = ( {id=0} = {}

) =>({
    type:"REMOVE_EXPENSE",
    id
})
