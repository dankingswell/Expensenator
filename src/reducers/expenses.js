// Expenses reducer
//takes default state and action

const expenseReducerDefaultState = []
const expensesReducer = (state = expenseReducerDefaultState,  action) =>{

    switch(action.type){
        case "ADD_EXPENSE":
        // ES6 SpreadOperator =  state.concat(action.expenses)
        // [...Arrayname, "new Prop"]  ... means take all vals from array mentioned      
        return [...state, action.expenses];
        
        case "REMOVE_EXPENSE":
            return state.filter((item)=>{
                return item.id != action.id   
            });
        
        case "EDIT_EXPENSE":
            return state.map((expense)=>{
                if (expense.id == action.id ){
                    return {
                        ...expense,
                        // as the functions passes an object it acts as a KVP
                        ...action.updates
                       }
                    }else{
                    return expense;
                }
            });

        default:
            return state;
    }

}

export default expensesReducer;