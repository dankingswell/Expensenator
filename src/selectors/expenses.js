import moment from "moment"

const getVisibileExpenses = (expenses,{text, sortBy ,startDate,endDate}) =>{
    // filter expense and if all fields match then return value
    return expenses.filter( (expense) => {
       
        const createdAtMoment = moment(expense.createdAt)
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, "day"):true;
        const endDateMatch = endDate  ?  endDate.isSameOrAfter(createdAtMoment, "day") : true;
        const textMatch = typeof(text) != "string" ||  expense.description.toLowerCase().includes(text.toLowerCase()) ;
        return startDateMatch && endDateMatch && textMatch;
    } )
    // sorting return data
    .sort( (a, b) => {
        if(sortBy == "date"){
            return a.createdAt < b.createdAt ? 1 : -1
        }
        if(sortBy == "amount"){
            return a.amount > b.amount ? -1 : 1;
        }
    });
}

export default getVisibileExpenses; 