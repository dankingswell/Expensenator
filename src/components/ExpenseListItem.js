import React from "react";
import {removeExpense} from "../actions/expenses";
import { Link } from "react-router-dom";
import moment from "moment";
import numeral from "numeral";


const ExpenseListItem = ({description, amount, createdAt,dispatch,id}) => (
    <div>
        <Link to={`/edit/${id}`} component={"./EditExpensePage.js"}>
            <h3>{description}</h3>
        </Link>

        <p>
        {`Â£${numeral(amount/100).format("0,0.00")}
           -  ${moment(createdAt).format("Do MMMM YY")}`}
        </p>
        
    </div>
    );

export default ExpenseListItem;