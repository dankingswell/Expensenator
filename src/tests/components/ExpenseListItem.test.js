import React from "react"
import {shallow} from "enzyme"
import ExpenseListItem from "../../components/ExpenseListItem"
import expenses from "../fixture/expenses"

test("item should render correctly wit all data", () => {
    const wrapper = shallow(<ExpenseListItem {...expenses[0]} />);
    expect(wrapper).toMatchSnapshot();
})