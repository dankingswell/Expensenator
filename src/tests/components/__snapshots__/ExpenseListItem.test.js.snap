// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`item should render correctly wit all data 1`] = `
<div>
  <Link
    component="./EditExpensePage.js"
    replace={false}
    to="/edit/1"
  >
    <h3>
      gas bill
    </h3>
  </Link>
  <p>
    Â£5.00
           -  1st January 70
  </p>
</div>
`;
